lib/temporal/client.rb

```diff
@@ -0,0 +1,382 @@
+        ),
+        signal_name: input.signal,
+        input: converter.to_payloads(input.args),
+        header: nil, # TODO: converter.to_payloads headers
+      )
+
+      connection.signal_workflow_execution(request)
+
+      return
+    end
+
+    def handle_cancel_workflow(input)
+      request = Temporal::Api::WorkflowService::V1::RequestCancelWorkflowExecutionRequest.new(
+        identity: identity,
+        request_id: SecureRandom.uuid,
+        namespace: namespace,
+        workflow_execution: Temporal::Api::Common::V1::WorkflowExecution.new(
+          workflow_id: input.id,
+          run_id: input.run_id || '',
+        ),
+        first_execution_run_id: input.first_execution_run_id || '',
+        reason: input.reason,
+      )
+
+      connection.request_cancel_workflow_execution(request)
+
+      return
+    end
+
+    def handle_terminate_workflow(input)
+      request = Temporal::Api::WorkflowService::V1::TerminateWorkflowExecutionRequest.new(
+        identity: identity,
+        namespace: namespace,
+        workflow_execution: Temporal::Api::Common::V1::WorkflowExecution.new(
+          workflow_id: input.id,
+          run_id: input.run_id || '',
+        ),
+        first_execution_run_id: input.first_execution_run_id || '',
+        reason: input.reason,
+        details: converter.to_payloads(input.args),
+      )
+
+      connection.terminate_workflow_execution(request)
+
+      return
+    end
+
+    def process_workflow_result_from(response)
+      if response.history.events.empty?
+        throw(:next) # next loop iteration
+      elsif response.history.events.length != 1
+        raise Temporal::Error, "Expected single close event, got #{response.history.events.length}"
+      end
+
+      event = response.history.events.first
+
+      case event.event_type
+      when :EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED
+        attributes = event.workflow_execution_completed_event_attributes
+        # TODO: Handle incorrect payloads object
+        return converter.from_payloads(attributes.result)&.first
+
+      when :EVENT_TYPE_WORKFLOW_EXECUTION_FAILED
+        # TODO: Use more specific error and decode failure
+        raise Temporal::Error, 'Workflow execution failed'
+
+      when :EVENT_TYPE_WORKFLOW_EXECUTION_CANCELED
+        # TODO: Use more specific error and decode failure
+        raise Temporal::Error, 'Workflow execution cancelled'
+
+      when :EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED
+        # TODO: Use more specific error and decode failure
+        raise Temporal::Error, 'Workflow execution terminated'
+
+      when :EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT
+        # TODO: Use more specific error and decode failure
+        raise Temporal::Error, 'Workflow execution timed out'
+
+      when :EVENT_TYPE_WORKFLOW_EXECUTION_CONTINUED_AS_NEW
+        # TODO: Use more specific error and decode failure
+        raise Temporal::Error, 'Workflow execution continued as new'
```

Comment:
We typically expose an option to follow the workflow chain and return the final result (the default behavior), or a user can explicitly opt in the getting the result for only the specific workflow they asked about, and only then should we return this error

It also isn't really an error state (even though we often trigger it when by raising errors in workflows, which I also dislike, but w/e)... so maybe we would consider returning a different type here rather than raising, but, I'll leave that up to whatever you think feels more natural.

Reply:
Yep, it's missing the `follow_runs` implementations. To be added as suggested above.

And I agree, raising an error here feels a bit wrong, but I believe this is how other SDKs are implementing it